File: Mobile/android/app/src/main/java/com/anonymous/FitConnect/GarminManager.kt
package com.anonymous.FitConnect

import android.app.Activity
import android.content.Context
import com.garmin.android.connectiq.ConnectIQ
import com.garmin.android.connectiq.ConnectIQ.ConnectIQListener
import com.garmin.android.connectiq.ConnectIQ.IQApplicationEventListener
import com.garmin.android.connectiq.ConnectIQ.IQDeviceEventListener
import com.garmin.android.connectiq.ConnectIQ.IQMessageStatus
import com.garmin.android.connectiq.ConnectIQ.IQSdkErrorStatus
import com.garmin.android.connectiq.ConnectIQ.IQConnectType
import com.garmin.android.connectiq.IQApp
import com.garmin.android.connectiq.IQDevice
import com.garmin.android.connectiq.exception.InvalidStateException
import com.garmin.android.connectiq.exception.ServiceUnavailableException

object GarminManager :
    ConnectIQListener,
    IQApplicationEventListener,
    IQDeviceEventListener {

    // Garmin's Android sample uses the manifest UUID as 32 hexadecimal
    // characters (without separators) for IQApp registration and messaging.
    private const val WATCH_APP_ID = "3051d135d05546578548018c798198a5"
    // A physical Android phone connected to the desktop watch simulator still
    // uses Garmin's ADB-backed TETHERED transport.
    private const val USE_TETHERED_SIMULATOR = true

    @Volatile
    var status: String = "GarminManager created"

    @Volatile
    var lastMessage: String = "No message yet"

    @Volatile
    var receivedMessageCount: Int = 0

    @Volatile
    var lastReceivedAt: Long = 0

    private var appContext: Context? = null
    private var connectIQ: ConnectIQ? = null
    private var selectedDevice: IQDevice? = null
    private var listenerRegistered = false

    private val connectType = if (USE_TETHERED_SIMULATOR) {
        IQConnectType.TETHERED
    } else {
        IQConnectType.WIRELESS
    }

    private val activeApp = IQApp(WATCH_APP_ID)

    fun initialize(activity: Activity) {
        appContext = activity.applicationContext
        status = "Initializing SDK"

        connectIQ = ConnectIQ.getInstance(
            appContext,
            connectType
        )

        try {
            connectIQ?.initialize(appContext, true, this)
            status = "SDK initialize called"
        } catch (e: Exception) {
            status = "SDK initialize exception"
            lastMessage = e.toString()
        }
    }

    override fun onSdkReady() {
        status = "SDK ready"
        refreshDevicesAndRegister()
    }

    override fun onInitializeError(error: IQSdkErrorStatus?) {
        status = "SDK init error: $error"
    }

    override fun onSdkShutDown() {
        status = "SDK shutdown"
        selectedDevice = null
        listenerRegistered = false
    }

    fun refreshDevicesAndRegister() {
        val ciq = connectIQ

        if (ciq == null) {
            status = "ConnectIQ is null"
            return
        }

        try {
            val knownDevices = ciq.knownDevices
            val connectedDevices = ciq.connectedDevices

            if (knownDevices.isEmpty()) {
                status = "No known Garmin devices"
                lastMessage = "Known=0, Connected=${connectedDevices.size}"
                return
            }

            val device = connectedDevices.firstOrNull() ?: knownDevices[0]
            selectedDevice = device

            lastMessage =
                "Selected device:\n" +
                "Name: ${device.friendlyName}\n" +
                "Known count: ${knownDevices.size}\n" +
                "Connected count: ${connectedDevices.size}"

            registerDeviceEvents(device)

            if (connectedDevices.any { it.deviceIdentifier == device.deviceIdentifier }) {
                registerAppListeners(device)
            } else {
                listenerRegistered = false
                status = if (USE_TETHERED_SIMULATOR) {
                    "Waiting for simulator connection"
                } else {
                    "Waiting for Garmin watch connection"
                }
            }

        } catch (e: InvalidStateException) {
            status = "Invalid state while reading devices"
            lastMessage = e.toString()
        } catch (e: ServiceUnavailableException) {
            status = "Garmin service unavailable"
            lastMessage = e.toString()
        } catch (e: Exception) {
            status = "Device refresh exception"
            lastMessage = e.toString()
        }
    }

    private fun registerDeviceEvents(device: IQDevice) {
        val ciq = connectIQ

        if (ciq == null) {
            status = "ConnectIQ is null during register"
            return
        }

        try {
            ciq.registerForDeviceEvents(device, this)
            status = "Device events registered: ${device.friendlyName}"
        } catch (e: Exception) {
            status = "Device event registration failed"
            lastMessage = e.toString()
            return
        }
    }

    private fun registerAppListeners(device: IQDevice) {
        val ciq = connectIQ

        if (ciq == null) {
            status = "ConnectIQ is null during app register"
            return
        }

        try {
            ciq.registerForAppEvents(device, activeApp, this)

            listenerRegistered = true
            status = if (USE_TETHERED_SIMULATOR) {
                "Simulator listener registered"
            } else {
                "Physical watch listener registered"
            }
            lastMessage = "Registered app id: ${activeApp.applicationId}"
                
        } catch (e: Exception) {
            status = "Garmin app listener registration failed"
            lastMessage = e.toString()
        }
    }

    override fun onDeviceStatusChanged(
        device: IQDevice?,
        deviceStatus: IQDevice.IQDeviceStatus?
    ) {
        status = "Device status: ${device?.friendlyName} = $deviceStatus"

        if (device != null && deviceStatus == IQDevice.IQDeviceStatus.CONNECTED) {
            selectedDevice = device
            registerAppListeners(device)
        } else if (deviceStatus != IQDevice.IQDeviceStatus.CONNECTED) {
            listenerRegistered = false
            selectedDevice = null
        }
    }

    override fun onMessageReceived(
        device: IQDevice?,
        app: IQApp?,
        message: MutableList<Any>?,
        messageStatus: IQMessageStatus?
    ) {
        status = "Message received: $messageStatus"

        if (message == null || message.isEmpty()) {
            lastMessage = "Empty message"
            return
        }

        receivedMessageCount += 1
        lastReceivedAt = System.currentTimeMillis()
        lastMessage =
            "Device: ${device?.friendlyName}\n" +
            "App: $app\n" +
            "Status: $messageStatus\n" +
            "Message:\n" +
            message.joinToString(separator = "\n\n") { item ->
                item.toString()
            }
    }

    fun getDebugStatus(): String {
        return status
    }

    fun getDebugMessage(): String {
        return lastMessage
    }

    fun isPhysicalDeviceMode(): Boolean {
        return !USE_TETHERED_SIMULATOR
    }

    fun sendPingToWatch() {
        val ciq = connectIQ
        val device = selectedDevice

        if (ciq == null || device == null) {
            status = "Cannot send: ciq/device null"
            return
        }

        status = "Sending Android ping"

        try {
            ciq.sendMessage(device, activeApp, "hello from Android") { _, _, sendStatus ->
                status = "Android send status: $sendStatus"
                lastMessage = "Android → Watch ping attempted"
            }
        } catch (e: Exception) {
            status = "Android send exception"
            lastMessage = e.toString()
        }
    }
}

File: Mobile/ios/GarminManager.swift
import Foundation
import ConnectIQ

final class GarminManager: NSObject, IQDeviceEventDelegate, IQAppMessageDelegate, IQUIOverrideDelegate {
    static let shared = GarminManager()

    private static let watchAppUUIDString = "3051d135-d055-4657-8548-018c798198a5"
    private static let returnURLScheme = "fitconnect"

    private var selectedDevice: IQDevice?
    private var activeApp: IQApp?
    private var listenerRegistered = false
    private var deviceReadyForCommunication = false

    private var ciq: ConnectIQ? {
        ConnectIQ.sharedInstance()
    }

    private(set) var status = "GarminManager created"
    private(set) var lastMessage = "No message yet"
    private(set) var receivedMessageCount = 0
    private(set) var lastReceivedAt: Int64 = 0

    private override init() {
        super.init()
    }

    func initialize() {
        guard let ciq else {
            status = "ConnectIQ singleton unavailable"
            return
        }

        ciq.initialize(
            withUrlScheme: Self.returnURLScheme,
            uiOverrideDelegate: self
        )

        status = "ConnectIQ initialized"
        lastMessage = "URL scheme: \(Self.returnURLScheme)"
    }

    func refreshDevicesAndRegister() {
        guard let selectedDevice else {
            requestDeviceSelection()
            return
        }

        registerDeviceEvents(for: selectedDevice)
    }

    func requestDeviceSelection() {
        guard let ciq else {
            status = "ConnectIQ singleton unavailable"
            return
        }

        status = "Opening Garmin device selection"
        ciq.showDeviceSelection()
    }

    @discardableResult
    func handleOpenURL(_ url: URL) -> Bool {
        guard url.scheme == Self.returnURLScheme else {
            return false
        }

        guard
            let ciq,
            let devices = ciq.parseDeviceSelectionResponse(from: url) as? [IQDevice],
            let device = devices.first
        else {
            status = "No Garmin devices returned"
            lastMessage = "Garmin Connect returned no compatible devices"
            return false
        }

        clearConnectionState(status: "Garmin device selected")

        selectedDevice = device
        lastMessage = """
        Selected device:
        Name: \(device.friendlyName ?? "Unknown")
        Model: \(device.modelName ?? "Unknown")
        Devices returned: \(devices.count)
        """

        registerDeviceEvents(for: device)
        return true
    }

    private func registerDeviceEvents(for device: IQDevice) {
        guard let ciq else {
            status = "ConnectIQ unavailable during device registration"
            return
        }

        ciq.unregister(forDeviceEvents: device, delegate: self)
        ciq.register(forDeviceEvents: device, delegate: self)

        selectedDevice = device

        let currentStatus = ciq.getDeviceStatus(device)

        status = "Device events registered; checking watch app"
        lastMessage = """
        Device: \(device.friendlyName ?? "Unknown")
        Current device status: \(currentStatus.rawValue)
        """

        registerAppMessages(for: device)
    }

    private func registerAppMessages(for device: IQDevice) {
        guard let ciq else {
            status = "ConnectIQ unavailable during app registration"
            return
        }

        guard let watchAppUUID = UUID(uuidString: Self.watchAppUUIDString) else {
            status = "Invalid watch app UUID"
            lastMessage = Self.watchAppUUIDString
            return
        }

        let app = IQApp(
            uuid: watchAppUUID,
            store: nil,
            device: device
        )

        activeApp = app
        listenerRegistered = false

        status = "Checking FitConnect app on watch"

        ciq.getAppStatus(app) { [weak self] appStatus in
            guard let self else { return }

            guard let appStatus else {
                self.activeApp = nil
                self.status = "App status lookup returned nil"
                self.lastMessage = "Could not query FitConnect on selected watch"
                return
            }

            self.lastMessage = "App status: \(String(describing: appStatus))"

            guard appStatus.isInstalled else {
                self.activeApp = nil
                self.listenerRegistered = false
                self.status = "FitConnect watch app is not installed"
                return
            }

            ciq.unregister(forAppMessages: app, delegate: self)
            ciq.register(forAppMessages: app, delegate: self)

            self.listenerRegistered = true
            self.status = "Physical watch listener registered"
            self.lastMessage = "Registered app UUID:\n\(Self.watchAppUUIDString)"
        }
    }

    func deviceStatusChanged(_ device: IQDevice!, status deviceStatus: IQDeviceStatus) {
        guard let device else {
            status = "Received device status without device"
            return
        }

        selectedDevice = device

        switch deviceStatus {
        case .connected:
            status = "Watch connected"
            lastMessage = "Waiting for characteristics discovery"

        case .bluetoothNotReady:
            clearConnectionState(status: "Bluetooth not ready")

        case .invalidDevice:
            clearConnectionState(status: "Invalid Garmin device")

        case .notConnected:
            clearConnectionState(status: "Watch disconnected")

        case .notFound:
            clearConnectionState(status: "Garmin watch not found")

        @unknown default:
            status = "Unknown device status"
            lastMessage = "Raw status: \(deviceStatus.rawValue)"
        }
    }

    func deviceCharacteristicsDiscovered(_ device: IQDevice!) {
        guard let device else {
            status = "Characteristics discovered without device"
            return
        }

        selectedDevice = device
        deviceReadyForCommunication = true
        status = "Watch ready for communication"

        registerAppMessages(for: device)
    }

    func receivedMessage(_ message: Any!, from app: IQApp!) {
        guard let message else {
            status = "Received empty watch message"
            lastMessage = "Empty message"
            return
        }

        receivedMessageCount += 1
        lastReceivedAt = Int64(Date().timeIntervalSince1970 * 1000)

        status = "Message received"
        lastMessage = """
        App: \(String(describing: app))
        Message:
        \(formatMessage(message))
        """
    }

    func sendPingToWatch() {
        guard let ciq else {
            status = "Cannot send: ConnectIQ unavailable"
            return
        }

        guard let app = activeApp else {
            status = "Cannot send: app not registered"
            return
        }

        guard deviceReadyForCommunication else {
            status = "Cannot send: watch not ready"
            return
        }

        guard listenerRegistered else {
            status = "Cannot send: listener not registered"
            return
        }

        let payload: NSDictionary = [
            "type": "ping",
            "source": "ios",
            "message": "hello from iPhone",
            "timestamp": NSNumber(value: Int64(Date().timeIntervalSince1970 * 1000))
        ]

        status = "Sending iPhone ping"

        ciq.sendMessage(
            payload,
            to: app,
            progress: nil
        ) { [weak self] result in
            self?.status = "iPhone send result: \(result)"
            self?.lastMessage = "iPhone → Watch payload attempted"
        }
    }

    func needsToInstallConnectMobile() {
        status = "Garmin Connect is required"
        lastMessage = "Install/open Garmin Connect and pair the Fenix."
    }

    func shutdown() {
        ciq?.unregister(forAllDeviceEvents: self)
        ciq?.unregister(forAllAppMessages: self)

        clearConnectionState(status: "Garmin SDK listeners removed")
    }

    func getDebugStatus() -> String {
        status
    }

    func getDebugMessage() -> String {
        lastMessage
    }

    func isPhysicalDeviceMode() -> Bool {
        true
    }

    private func clearConnectionState(status newStatus: String) {
        selectedDevice = nil
        activeApp = nil
        listenerRegistered = false
        deviceReadyForCommunication = false
        status = newStatus
    }

    private func formatMessage(_ message: Any) -> String {
        guard JSONSerialization.isValidJSONObject(message) else {
            return String(describing: message)
        }

        guard
            let data = try? JSONSerialization.data(
                withJSONObject: message,
                options: [.prettyPrinted, .sortedKeys]
            ),
            let text = String(data: data, encoding: .utf8)
        else {
            return String(describing: message)
        }

        return text
    }
}
